pkg/querier/tail.go

```diff
@@ -96,44 +101,62 @@ func (t *Tailer) loop() {
-				if len(t.querierTailClients) == 0 {
-					// All the connections to ingesters are dropped, try reconnecting or return error
-					if err := t.checkIngesterConnections(); err != nil {
-						level.Error(util.Logger).Log("Error reconnecting to ingesters", fmt.Sprintf("%v", err))
-					} else {
-						continue
-					}
-					if err := t.close(); err != nil {
-						level.Error(util.Logger).Log("Error closing Tailer", fmt.Sprintf("%v", err))
-					}
-					t.closeErrChan <- errors.New("all ingesters closed the connection")
-					return
-				}
-				time.Sleep(nextEntryWait)
-				continue
-			}
-		} else {
-			// If channel is blocked already, drop current entry directly to save the effort
-			if t.isBlocked() {
-				t.dropEntry(t.currEntry.Timestamp, t.currLabels, nil)
+		// Read as much entries as we can (up to the max allowed) and populate the
+		// tail response we'll send over the response channel
+		tailResponse := new(TailResponse)
+		entriesCount := 0
+
+		for ; entriesCount < maxEntriesPerTailResponse && t.next(); entriesCount++ {
+			// If the response channel channel is blocked, we drop the current entry directly
+			// to save the effort
+			if t.isResponseChanBlocked() {
+				droppedEntries = dropEntry(droppedEntries, t.currEntry.Timestamp, t.currLabels)
 				continue
 			}
 
-			tailResponse.Streams = append(tailResponse.Streams, logproto.Stream{Labels: t.currLabels, Entries: []logproto.Entry{t.currEntry}})
-			if len(tailResponse.Streams) != 100 {
-				continue
+			tailResponse.Streams = append(tailResponse.Streams, logproto.Stream{
+				Labels:  t.currLabels,
+				Entries: []logproto.Entry{t.currEntry},
+			})
+		}
+
+		// If all consumed entries have been dropped because the response channel is blocked
+		// we should reiterate on the loop
+		if len(tailResponse.Streams) == 0 && entriesCount > 0 {
+			continue
+		}
+
+		// If no entry has been consumed we should ensure it's not caused by all ingesters
+		// connections dropped and then throttle for a while
+		if len(tailResponse.Streams) == 0 {
+			if len(t.querierTailClients) == 0 {
+				// All the connections to ingesters are dropped, try reconnecting or return error
+				if err := t.checkIngesterConnections(); err != nil {
+					level.Error(util.Logger).Log("Error reconnecting to ingesters", fmt.Sprintf("%v", err))
+				} else {
+					continue
+				}
+				if err := t.close(); err != nil {
+					level.Error(util.Logger).Log("Error closing Tailer", fmt.Sprintf("%v", err))
+				}
+				t.closeErrChan <- errors.New("all ingesters closed the connection")
+				return
 			}
-			tailResponse.DroppedEntries = t.popDroppedEntries()
+
+			time.Sleep(t.waitEntryThrottle)
+			continue
 		}
 
+		// Send the tail response through the response channel without blocking.
+		// Drop the entry if the response channel buffer is full.
+		tailResponse.DroppedEntries = droppedEntries
+
 		select {
 		case t.responseChan <- tailResponse:
+			droppedEntries = make([]droppedEntry, 0)
```

Comment:
Lets put a check here for the size of droppedEntries before re-initializing?
If we are doing this, we would also want to put same condition above where droppedEntries is assigned to tailResponse i.e https://github.com/grafana/loki/pull/877/files#diff-6786afcc7f6bd0a7f048371a46467af0R152

Reply:
The way I read your feedback is "do it only `if len(droppedEntries) > 0`". If I correctly understood your comment, then definitely yes (done). If I misunderstood your comment, may you share more details on your thought, please?

Reply:
That is what I meant, thanks for fixing this!